### PR TITLE
defer/stream: split incremental delivery into new entry points

### PR DESCRIPTION
**Pull request by glasser**
This PR changes the `execute`, `subscribe`, and `graphql` APIs on the
`defer-stream` branch back to having the same API as on `main`: they can only
produce a single `ExecutionResult` (or a stream of `ExecutionResult`s for
`subscribe`) which never has a `hasNext` field, and do not support incremental
delivery.

Incremental delivery is now provided by the new entry points
`experimentalExecuteIncrementally` and `experimentalSubscribeIncrementally`. (We
will remove "experimental" once the proposal has been merged into the GraphQL
spec.)

This will make upgrading to graphql@17 easier, and increases the clarity
of return types.

Use distinct types for "the only result of a single-payload execution",
"the first payload of a multi-payload execution", and "subsequent
payloads of a multi-payload execution", since the data structures are
different. (Multi-payload executions *always* have at least one payload,
and the structure differs, which is why the new types separate the
initial result from subsequent results.)

Namely, single results have no `hasNext` field; initial results have `hasNext`
and may combine `data`/`errors` with `incremental`; and subsequent results have
`hasNext` and `incremental`.

Note that with the previous types, you actually had to use a function
with a type guard like `isAsyncIterable`: you couldn't just write `if
(Symbol.asyncIterator in result)`. I think both explicitly separating
the first (differently-typed) element from the rest of the elements
*and* making it possible to differentiate the types with a simple `in`
check are improvements. Additionally, somebody using the API in
JavaScript who doesn't realize that there is new functionality available
who is confused that the result no longer has `data`/`errors` fields can
just print out the result and see and search for the terms
`initialResult` and `subsequentResults` instead of only seeing iterator
internals.

Fix `Formatted*Result` types to use `GraphQLFormattedError` rather than
`GraphQLError` for `errors` nested inside `incremental`.

### CODE REVIEW

**Comment by glasser** [comment]
Updated this PR to be less of an experiment and more of an actual proposal. Updated the PR description too.

**Comment by yaacovCR** [comment]
This seems to have another tiny advantage besides typings: If we return a generator that includes/emits the first payload (current behavior) rather than returning the first payload separately (as in this PR), then even if the first payload is produced synchronously, the client will have to await it.

Even if it’s produced asynchronously, with current behavior, the client will have to await the generator (which is not produced until the first payload is) and then again have to await the first payload emitted by the generator.

Perhaps this pattern also makes sense generally for every generator that is known to be non-empty…

**Comment by glasser** [comment]
@robrichard @IvanGoncharov I've rebased this on top of #3659 with the #3710 fix incorporated, and I've updated the PR description to be more informative and less "experimental".

**Comment by glasser** [comment]
After discussion with @IvanGoncharov I've updated this to move incremental delivery out of `execute` into a new entry point `experimentalExecuteIncrementally` (which I recognize is not quite aligned with the discussion at https://github.com/robrichard/defer-stream-wg/discussions/12). This seems like the most appropriate way to let us move forward.

File: src/execution/execute.ts [hunk 1]
```diff
@@ -210,12 +268,58 @@ export interface ExecutionArgs {
...
-): PromiseOrValue<
-  ExecutionResult | AsyncGenerator<AsyncExecutionResult, void, void>
-> {
+): Promise<ExperimentalExecuteIncrementallyResults> {
+  const result = executeSyncOrIncrementally(args);
+  if (isPromise(result)) {
+    return result;
+  }
+  // Always return a Promise for a consistent API.
+  return Promise.resolve({ singleResult: result });
```
Comment (yaacovCR) [inline src/execution/execute.ts:313]: Question: Is the benefit of always returning a promise definitely worth it?

1. If the first result is synchronous, 
returning the generator could also be synchronous and then we save a tick.

2. This change actually sort of makes the API inconsistent in another sense as non incremental delivery can be either value or Promise, so maybe shouldn't incremental.

3. I believe we recently changed in v17 subscribe to allow returning a generator synchronously. Looks like now we r headed in the other direction. Either way, should they be aligned?

File: src/execution/execute.ts [hunk 2]
```diff
@@ -210,12 +268,58 @@ export interface ExecutionArgs {
...
+}
+
+/**
+ * Helper for the other execute functions. Returns an ExecutionResult
+ * synchronously if all resolvers return non-Promises
+ */
+function executeSyncOrIncrementally(
```
Comment (yaacovCR) [inline src/execution/execute.ts:320]: Do we need this combination?

**Comment by glasser** [comment]
@github-actions publish-pr-on-npm

File: src/graphql.ts [hunk 3]
```diff
@@ -71,9 +69,7 @@ export interface GraphQLArgs {
   typeResolver?: Maybe<GraphQLTypeResolver<any, any>>;
 }
 
-export function graphql(
-  args: GraphQLArgs,
-): Promise<ExecutionResult | AsyncGenerator<AsyncExecutionResult, void, void>> {
+export function graphql(args: GraphQLArgs): Promise<ExecutionResult> {
```
[the remark this answers is not in the document]
Comment (glasser) [reply src/graphql.ts:72]: I think this was @IvanGoncharov 's idea but I might be misinterpreting. Probably planning to change when not "experimental"?

File: src/execution/execute.ts [hunk 4]
```diff
@@ -150,44 +148,104 @@ export interface FormattedExecutionResult<
...
+        SubsequentIncrementalExecutionResult<TData, TExtensions>,
+        void,
+        void
+      >;
+    };
+
+export interface InitialIncrementalExecutionResult<
+  TData = ObjMap<unknown>,
+  TExtensions = ObjMap<unknown>,
+> extends ExecutionResult<TData, TExtensions> {
```
[the remark this answers is not in the document]
Comment (glasser) [reply src/execution/execute.ts:171]: Fixed.

**Comment by glasser** [comment]
@github-actions publish-pr-on-npm

**Comment by glasser** [comment]
I should apply the same "un-transformation" to `subscribe`: split it into a v16-esque `subscribe` and `experimentalSubscribeIncrementally`.

File: src/graphql.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (IvanGoncharov) [reply src/graphql.ts]: @robrichard Yes, @glasser is right.
The idea is to finally merge it into `main` having an "experiment" prefix to everything.
Changing the prototype of `graphql` makes it a breaking change, so better to avoid it before we are sure that the shape of the response is final.

So this PR has two things inside:
1. Stuff @glasser learned by implementing this feature in ApolloServer (changes to `execute` typings).
2. Moving everything under the "experiment" prefix and cutting non-essential functionality (e.g. `graphql`).

File: src/execution/execute.ts (hunk 2 above)
Comment (IvanGoncharov) [reply src/execution/execute.ts:326]: I agree, it with @yaacovCR on that. It's DRY and saves a few lines. But given all the planned changes to `execute` API, it would be better to just duplicate those lines in both functions.

File: src/execution/execute.ts (hunk 1 above)
Comment (IvanGoncharov) [reply src/execution/execute.ts:319]: @yaacovCR  Agree, with can make this function return `PromiseOrValue<ExperimentalExecuteIncrementallyResults>`

File: src/execution/execute.ts (hunk 2 above)
Comment (IvanGoncharov) [reply src/execution/execute.ts:326]: @yaacovCR Also, it's a non-exported function, just a utility one to be DRY.
But I still think it's somewhat confusing and don't worth the DRY benefits it gives.

File: src/execution/__tests__/simplePubSub.ts [hunk 5]
```diff
@@ -51,6 +51,7 @@ export class SimplePubSub<T> {
         emptyQueue();
         return Promise.resolve({ value: undefined, done: true });
       },
+      /* c8 ignore next 4 */
```
Comment (robrichard) [inline src/execution/__tests__/simplePubSub.ts:54]: why do we need to ignore these?

File: src/execution/__tests__/lists-test.ts [hunk 6]
```diff
@@ -113,7 +113,7 @@ describe('Execute: Accepts async iterables as list value', () => {
         },
       }),
     });
-    return execute({
+    return experimentalExecuteIncrementally({
```
Comment (robrichard) [inline src/execution/__tests__/lists-test.ts:116]: The tests in this file don't use defer or stream. Why did you change it to use `experimentalExecuteIncrementally`?

**Review by robrichard — APPROVED** [review]

File: src/execution/__tests__/simplePubSub.ts (hunk 5 above)
Comment (glasser) [reply src/execution/__tests__/simplePubSub.ts:54]: Good call. We don't any more! I will check any other additional c8 ignores I added.

File: src/execution/__tests__/lists-test.ts (hunk 6 above)
Comment (glasser) [reply src/execution/__tests__/lists-test.ts:118]: Good catch. It doesn't need to be, and my PR will now remove the `| AsyncGenerator` from `completeObjectList`'s return type, making it even more clear that the tests in this file don't use defer or stream.

**Comment by glasser** [comment]
Today's updates:
- As suggested by @yaacovCR, `experimentalExecuteIncrementally` now returns PromiseOrValue.
- Treat subscribe the same way as I treated execute: `subscribe` does the same thing as in v6 (with errors in the stream if you use `@defer`/`@stream`), and `experimentalSubscribeIncrementally` supports incremental delivery.
- As per spec, initial results may have `incremental` fields, and all top-level results may have `extensions`.
- Testing improvements from @robrichard.

**Comment by glasser** [comment]
@github-actions publish-pr-on-npm
